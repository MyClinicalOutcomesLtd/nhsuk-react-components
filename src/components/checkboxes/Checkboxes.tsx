import classNames from 'classnames';
import React from 'react';
import useCheckboxes from '../../util/hooks/UseCheckboxes';
import useFormGroup from '../../util/hooks/UseFormGroup';
import { FormElementProps, FormGroupConsumer } from '../../util/types/FormTypes';
import CheckboxContext, { ICheckboxContext } from './CheckboxContext';
import Box from './components/Box';

interface CheckboxesProps extends FormElementProps<HTMLDivElement> {
  idPrefix?: string;
}

type CheckboxesChildElements = {
  Box: typeof Box;
};

const Checkboxes: React.FC<CheckboxesProps> & CheckboxesChildElements = (props) => {
  const group = useFormGroup(FormGroupConsumer.CHECKBOXES, props);
  const { conditionalBoxes, resetBoxIds, ...checkboxFuncs } = useCheckboxes(
    group.renderProps.id,
    group.renderProps.idPrefix,
  );

  // "error" prop is pulled out here so it is not passed to the div element
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const { className, name, children, error, ...rest } = group.renderProps;

  const containsConditional = conditionalBoxes.length > 0;
  const contextValue: ICheckboxContext = { name, ...checkboxFuncs };

  resetBoxIds();

  return (
    <group.Wrapper {...group.wrapperProps}>
      {group.LabelBlock}
      <div
        className={classNames(
          'nhsuk-checkboxes',
          { 'nhsuk-checkboxes--conditional': containsConditional },
          className,
        )}
        {...rest}
      >
        <CheckboxContext.Provider value={contextValue}>{children}</CheckboxContext.Provider>
      </div>
    </group.Wrapper>
  );
};

Checkboxes.Box = Box;

export default Checkboxes;
