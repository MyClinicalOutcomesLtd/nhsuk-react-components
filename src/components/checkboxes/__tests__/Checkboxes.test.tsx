import { mount } from 'enzyme';
import React from 'react';
import Checkboxes from '../Checkboxes';

describe('Checkboxes', () => {
  it('renders without errors', () => {
    const wrapper = mount(
      <Checkboxes id="test-id">
        <Checkboxes.Box value="british">British</Checkboxes.Box>
        <Checkboxes.Box value="irish">Irish</Checkboxes.Box>
        <Checkboxes.Box value="other">Citizen of another country</Checkboxes.Box>
      </Checkboxes>,
    );

    expect(wrapper).toMatchSnapshot();

    const allBoxes = wrapper.find(Checkboxes.Box);

    expect(allBoxes).toHaveLength(3);

    const firstBoxInput = allBoxes.at(0).find('input');
    expect(firstBoxInput.exists()).toBe(true);
    expect(firstBoxInput.prop('id')).toBe('test-id-1');
    expect(firstBoxInput.prop('name')).toBe('test-id');
    expect(firstBoxInput.prop('type')).toBe('checkbox');
    expect(firstBoxInput.prop('value')).toBe('british');

    const secondBoxInput = allBoxes.at(1).find('input');
    expect(secondBoxInput.exists()).toBe(true);
    expect(secondBoxInput.prop('id')).toBe('test-id-2');
    expect(secondBoxInput.prop('name')).toBe('test-id');
    expect(secondBoxInput.prop('type')).toBe('checkbox');
    expect(secondBoxInput.prop('value')).toBe('irish');

    const thirdBoxInput = allBoxes.at(2).find('input');
    expect(thirdBoxInput.exists()).toBe(true);
    expect(thirdBoxInput.prop('id')).toBe('test-id-3');
    expect(thirdBoxInput.prop('name')).toBe('test-id');
    expect(thirdBoxInput.prop('type')).toBe('checkbox');
    expect(thirdBoxInput.prop('value')).toBe('other');

    wrapper.unmount();
  });

  it('adds conditional class when a box contains conditional content', () => {
    const wrapper = mount(
      <Checkboxes id="test-id">
        <Checkboxes.Box conditional={<>Conditional Content</>}>British</Checkboxes.Box>
      </Checkboxes>,
    );

    const checkboxesEl = wrapper.find('div.nhsuk-checkboxes');
    expect(checkboxesEl.hasClass('nhsuk-checkboxes--conditional')).toBe(true);

    wrapper.unmount();
  });
});
