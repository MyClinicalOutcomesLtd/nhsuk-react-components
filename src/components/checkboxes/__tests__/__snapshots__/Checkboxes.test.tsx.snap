// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkboxes renders without errors 1`] = `
<Checkboxes
  id="test-id"
>
  <FormGroup
    _exposeContext={false}
    error={false}
  >
    <div
      className="nhsuk-form-group"
    >
      <div
        className="nhsuk-checkboxes"
        id="test-id"
      >
        <Box
          type="checkbox"
          value="british"
        >
          <div
            className="nhsuk-checkboxes__item"
          >
            <input
              className="nhsuk-checkboxes__input"
              id="test-id-1"
              name="test-id"
              onChange={[Function]}
              type="checkbox"
              value="british"
            />
            <Label
              className="nhsuk-checkboxes__label"
              htmlFor="test-id-1"
              id="test-id-1--label"
            >
              <BaseLabel
                className="nhsuk-checkboxes__label"
                htmlFor="test-id-1"
                id="test-id-1--label"
              >
                <label
                  className="nhsuk-label nhsuk-checkboxes__label"
                  htmlFor="test-id-1"
                  id="test-id-1--label"
                >
                  British
                </label>
              </BaseLabel>
            </Label>
          </div>
        </Box>
        <Box
          type="checkbox"
          value="irish"
        >
          <div
            className="nhsuk-checkboxes__item"
          >
            <input
              className="nhsuk-checkboxes__input"
              id="test-id-2"
              name="test-id"
              onChange={[Function]}
              type="checkbox"
              value="irish"
            />
            <Label
              className="nhsuk-checkboxes__label"
              htmlFor="test-id-2"
              id="test-id-2--label"
            >
              <BaseLabel
                className="nhsuk-checkboxes__label"
                htmlFor="test-id-2"
                id="test-id-2--label"
              >
                <label
                  className="nhsuk-label nhsuk-checkboxes__label"
                  htmlFor="test-id-2"
                  id="test-id-2--label"
                >
                  Irish
                </label>
              </BaseLabel>
            </Label>
          </div>
        </Box>
        <Box
          type="checkbox"
          value="other"
        >
          <div
            className="nhsuk-checkboxes__item"
          >
            <input
              className="nhsuk-checkboxes__input"
              id="test-id-3"
              name="test-id"
              onChange={[Function]}
              type="checkbox"
              value="other"
            />
            <Label
              className="nhsuk-checkboxes__label"
              htmlFor="test-id-3"
              id="test-id-3--label"
            >
              <BaseLabel
                className="nhsuk-checkboxes__label"
                htmlFor="test-id-3"
                id="test-id-3--label"
              >
                <label
                  className="nhsuk-label nhsuk-checkboxes__label"
                  htmlFor="test-id-3"
                  id="test-id-3--label"
                >
                  Citizen of another country
                </label>
              </BaseLabel>
            </Label>
          </div>
        </Box>
      </div>
    </div>
  </FormGroup>
</Checkboxes>
`;
